### PR TITLE
SpillSequenceSpiller – preserve `ref` BoundConditionalOperator in presence of spilling in its either branch

### LINKED ISSUE #74115: Bug: Conditional Operator stores ref result in a non-ref-temporary (Only when SpillSequenceSpiller is used - e.g. for async/await)
**Version Used**: [847d588](https://github.com/dotnet/roslyn/commit/847d58843a139405d633a55c82b232582c996848)

**Steps to Reproduce**:

Compile and execute the following [code](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWACgsAGAAizwFYBuYgfXIF5yAxAew4DoA5VMMACmyAEKoA5hICeACgCU3AEpCAzqgA2CeiTwBOWQCIAtKbOnD8nU1ace/QSN4cE4qXKvFiWAMyUc7FwA3sTkYWG+lHgAbJQAHJTRADywCAB85A7CYpIyCqHh5CFEhYVUFKoIyCzkhhA4hjql4ankAB41eE3N5K3SNTg6Bc3IQgBmvTAI5NXMwz2FoxNYeuQAogBuEBoAgqrSMADGspXIcOR48r2qtRB4huQA/DPj7fMLIC/Lq5vbewfHU7nHBXKA3OoNJ5fcjSd49T5LKLKdAIKBgITcADCHDAAAcoBoRABlEQbKCHNTcACqMFUEDGGP4Gg0KjGKSmaQU3WacPCyAA1PzuSNBcLSgKhbywq02rKasgxWU8NwqAZZW1PCUFlQVfpZBrFeEdarZNIrOQAPQW1QACw4mgAJuRhORqOQYBwAO4AQillAA7O1ZcKAL7zeaRKixABq21QQiwyVSGSyThcbjy8nexW1eAqVRqEMaUplnUNk2m/VYgy8WpGr1asz94tev12+yOJyq50uyjUmmmYNu9yhiLazbCCNbW3bAK7Z3IIL76i011ukOeiNhdYW5Cny2VShRaIx2LxBOJpPJlJpdIZfE0LPG7PSXKlfol5Zmoo/P53pRlOVWAVP1jT1dVNV3MC1Ug3NdQMM0vywQMYCET19VlWCwjDHco3IWMNHjRMknKDI23+TtynIVRzlaKAsx3HMemg2QABJDCCVQQyKKAQ0sctkPwuME2ibg2GQHEVBXBATiw8gcJwoA):

```csharp
using System;
using System.Threading.Tasks;
_ = Foo.NumberBuggy().Result;
Console.WriteLine("--------");
_ = Foo.NumberNotBuggy();

public class Foo{
     public static async Task<int> NumberBuggy()
     {
        string str = "a2";
        int x = 1;
        int y = 2;

        ref int r =
              ref await EvalAsync(str, 1) is "a1" ? ref x
            : ref await EvalAsync(str, 2) is "a2" ? ref y
            : ref System.Runtime.CompilerServices.Unsafe.NullRef<int>();
        
         r++;
         r++;
         r++;
         int xxx = r;
         System.Console.WriteLine(xxx);
         System.Console.WriteLine(x);
         System.Console.WriteLine(y); //should be 5 now!
         return xxx;
    }
    
    public static ValueTask<int> NumberNotBuggy()
     {
        string str = "a2";
        int x = 1;
        int y = 2;

        ref int r =
              ref EvalAsync(str, 1).Result is "a1" ? ref x
            : ref EvalAsync(str, 2).Result is "a2" ? ref y
            : ref System.Runtime.CompilerServices.Unsafe.NullRef<int>();
        
         r++;
         r++;
         r++;
         int xxx = r;
         System.Console.WriteLine(xxx);
         System.Console.WriteLine(x);
         System.Console.WriteLine(y);
         return new(xxx);
    }
    static ValueTask<string> EvalAsync(string s, int i)
    {
        System.Console.WriteLine($"{s} {i}");
        return ValueTask.FromResult(s);
    }
}
```

**Expected Behavior**:
Same output for both methods (they are not really async)

**Actual Behavior**:
The result of the conditional expression is stored in a non-ref temp local which causes that the later increment operations do not change the original value - just the temp.

**Error cause**
The error is caused in [SpillSequenceSpiller](https://github.com/dotnet/roslyn/blob/847d58843a139405d633a55c82b232582c996848/src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs#L1135
)

The code should be:
```csharp
var tmp = _F.SynthesizedLocal(node.Type, kind: SynthesizedLocalKind.Spill, syntax: _F.Syntax,
   refKind: node.IsRef ? RefKind.Ref : RefKind.None);
```

But how can we determine if the result should be ref-readonly or not?

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #74115.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefConditionalOperatorTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefConditionalOperatorTests.cs:1]: Consider adding a test case from the bug, without `await`, as that will produce simpler IL and may be easier to diagnose if we were to regress this area in the future.

**Review by 333fred — APPROVED** [review]
Generally LGTM, but I would suggest adding one of the test cases from the original bug as well.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

File: src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs [hunk 1]
```diff
@@ -1142,6 +1142,33 @@ public override BoundNode VisitConditionalOperator(BoundConditionalOperator node
...
+
+                conditionBuilder.AddStatement(
+                    _F.If(condition,
+                        _F.StatementList(consequenceBuilder.GetStatements()),
+                        _F.StatementList(alternativeBuilder.GetStatements())));
+
+                consequenceBuilder.Free();
+                alternativeBuilder.Free();
+
+                // 3. Use updated conditional operator as the result. Note, we are using the captured temp as its condition,
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs:1168]: >Use updated conditional operator as the result [](http://example.com/codeflow?start=22&length=46)

nit: consider calling out that we want to use the original node to maintain proper refness

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 3)

**Pull request merged**